### PR TITLE
Ck/18430 hoisting text attributes

### LINKED ISSUE #18430: Discrepancy in hoisting text attributes to blocks (to selection attributes)
## 📝 Provide a description of the improvement

https://github.com/user-attachments/assets/df8e9305-9519-4d9d-9c4e-34c95c17add8

Typing over:

```
<p><b>foo</b></p>
<p>[]</p>
<p><b>foo</b></p>
```

will behave differently depending on how this content was generated.

If it was generated by applying bold to the first paragraph and then using Enter key to create two more, the second one, even when empty, will preserve this attribute in a form of `selection:bold` on `paragraph`.

However, if bold is applied to 3 paragraphs `<p>foo</p><p></p><p>foo</p>` then the second one won't be styled.

## 📃 Other details

* Browser: …
* OS: …
* CKEditor version: …
* Installed CKEditor plugins: …

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by arkflpc**
### 🚀 Summary

Fixed a discrepancy where applying a text attribute (e.g. bold) to a selection that included empty paragraphs did not set stored selection attributes on those paragraphs.

---

### 📌 Related issues

* Closes #18430

---

### 💡 Additional information

*Optional: Notes on decisions, edge cases, or anything helpful for reviewers.*

---

### 🧾 Checklists

Use the following checklists to ensure important areas were not overlooked.
This does not apply to feature-branch merges.
If an item is **not relevant** to this type of change, simply leave it unchecked.

#### Author checklist

- [ ] Is the changelog entry intentionally omitted?
- [x] Is the change backward-compatible?
- [x] Have you considered the impact on different editor setups and core interactions? _(e.g., classic/inline/multi-root/many editors, typing, selection, paste, tables, lists, images, collaboration, pagination)_
- [x] Has the change been manually verified in the relevant setups?
- [ ] Does this change affect any of the above?
- [ ] Is performance impacted?
- [ ] Is accessibility affected?
- [ ] Have tests been added that fail without this change (against regression)?
- [ ] Have the API documentation, guides, feature digest, and related feature sections been updated where needed?
- [ ] Have metadata files (ckeditor5-metadata.json) been updated if needed?
- [ ] Are there any changes the team should be informed about (e.g. architectural, difficult to revert in future versions or having impact on other features)?
- [ ] Were these changes documented (in Logbook)?

#### Reviewer checklist

- [ ] PR description explains the changes and the chosen approach (especially when  performance, API, or UX is affected).
- [ ] The changelog entry is clear, user‑ or integrator-facing, and it describes any breaking changes.
- [ ] All new external dependencies have been approved and mentioned in LICENSE.md (if any).
- [ ] All human-readable, translateable strings in this PR been introduced using `t()` (if any).
- [ ] I manually verified the change (e.g., in manual tests or documentation).
- [ ] The target branch is correct.

---

> [!NOTE]
> **Medium Risk**
> Touches core schema range validation and changes how multiple formatting commands apply attributes, which could affect formatting behavior in other selection edge cases; coverage is improved with targeted regression tests.
> 
> **Overview**
> Fixes a formatting edge case where applying text attributes over a selection that spans empty blocks would not persist *stored selection attributes* for those empty blocks, causing the attribute state to be lost when the caret moved there.
> 
> This extends `ModelSchema#getValidRanges()` with an `includeEmptyRanges` option and updates `AttributeCommand`, `FontCommand`, `HighlightCommand`, and `TextPartLanguageCommand` to opt in, applying `selection:*` (stored) attributes when a returned valid range is collapsed. Regression tests were added across these features, and affected packages now list `@ckeditor/ckeditor5-engine` as a runtime dependency.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5925256ca6ad2ff0981e24ccf15b0ce158ebd81d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/ckeditor5-basic-styles/src/attributecommand.ts [hunk 1]
```diff
@@ -92,13 +93,24 @@ export class AttributeCommand extends Command {
...
 
 				for ( const range of ranges ) {
+					let itemOrRange: ModelRange | ModelElement = range;
+					let attributeKey = this.attributeKey;
+					const containedElement = range.getContainedElement();
+
+					if ( containedElement ) {
+						itemOrRange = containedElement;
+						attributeKey = `selection:${ this.attributeKey }`;
+					}
```
Comment (niegowski) [inline packages/ckeditor5-basic-styles/src/attributecommand.ts:107]: From the Schema point of view, it should return only valid ranges for the given attribute. The selected element does not allow such an attribute. But what if the `getValidRanges()` would return collapsed ranges inside empty elements that allow `$text` with the specified attribute? WDYT?

```suggestion
					let itemOrRange: ModelRange | ModelElement = range;
					let attributeKey = this.attributeKey;

					if ( range.isCollapsed ) {
						itemOrRange = range.start.parent;
						attributeKey = `selection:${ this.attributeKey }`;
					}
```

File: packages/ckeditor5-engine/src/model/schema.ts [hunk 2]
```diff
@@ -1156,13 +1163,25 @@ export class ModelSchema extends /* #__PURE__ */ ObservableMixin() {
...
+		options: { includeEmptyElements?: boolean } = {}
+	): Iterable<ModelRange> {
 		let start = range.start;
 		let end = range.start;
 
 		for ( const item of range.getItems( { shallow: true } ) ) {
 			if ( item.is( 'element' ) ) {
-				yield* this._getValidRangesForRange( ModelRange._createIn( item ), attribute );
+				if ( options.includeEmptyElements && item.isEmpty ) {
+					const context = [ ...item.getAncestors( { includeSelf: true } ) as any, new ModelText() ];
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/schema.ts]:
```suggestion
					// In empty element check if it accepts text with the given attribute.
					const context = this.createContext( item ).push( '$text' );
```

File: packages/ckeditor5-engine/src/model/schema.ts [hunk 3]
```diff
@@ -1156,13 +1163,25 @@ export class ModelSchema extends /* #__PURE__ */ ObservableMixin() {
 	 * @param attribute The name of the attribute to check.
 	 * @returns Ranges in which the attribute is allowed.
 	 */
-	private* _getValidRangesForRange( range: ModelRange, attribute: string ): Iterable<ModelRange> {
+	private* _getValidRangesForRange(
+		range: ModelRange,
+		attribute: string,
+		options: { includeEmptyElements?: boolean } = {}
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/schema.ts]: If we change the method output to a collapsed range, then suchan  option would fit better:

```suggestion
		options: { includeEmptyRanges?: boolean } = {}
```

File: packages/ckeditor5-engine/src/model/schema.ts [hunk 4]
```diff
@@ -1156,13 +1163,25 @@ export class ModelSchema extends /* #__PURE__ */ ObservableMixin() {
...
 		let end = range.start;
 
 		for ( const item of range.getItems( { shallow: true } ) ) {
 			if ( item.is( 'element' ) ) {
-				yield* this._getValidRangesForRange( ModelRange._createIn( item ), attribute );
+				if ( options.includeEmptyElements && item.isEmpty ) {
+					const context = [ ...item.getAncestors( { includeSelf: true } ) as any, new ModelText() ];
+
+					if ( this.checkAttribute( context, attribute ) ) {
+						yield ModelRange._createOn( item );
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/schema.ts]:
```suggestion
						yield ModelRange._createIn( item );
```

**Comment by niegowski** [comment]
Apart from the above, this PR includes only the basic-styles package. Will other text attributes (like font) be handled in a separate PR or added later to this PR?

**Comment by arkflpc** [comment]
> Apart from the above, this PR includes only the basic-styles package. Will other text attributes (like font) be handled in a separate PR or added later to this PR?

Good point. I'll add it in this PR.

File: packages/ckeditor5-language/src/textpartlanguagecommand.ts [hunk 5]
```diff
@@ -85,13 +86,23 @@ export class TextPartLanguageCommand extends Command {
...
+					includeEmptyRanges: true
+				} );
 
 				for ( const range of ranges ) {
+					let itemOrRange: ModelRange | ModelElement = range;
+					let attributeKey = 'language';
+
+					if ( range.isCollapsed ) {
+						itemOrRange = range.start.parent as ModelElement;
+						attributeKey = 'selection:language';
```
Comment (niegowski) [inline packages/ckeditor5-language/src/textpartlanguagecommand.ts]: I'd suggest using a dedicated method to construct attribute key:

https://github.com/ckeditor/ckeditor5/blob/b699e0768f213a12705b203d2aa4ae6f13970c7b/packages/ckeditor5-engine/src/model/documentselection.ts#L480-L482

It is marked as internal but maybe we could promote it to the public API? Together with:

https://github.com/ckeditor/ckeditor5/blob/b699e0768f213a12705b203d2aa4ae6f13970c7b/packages/ckeditor5-engine/src/model/documentselection.ts#L489-L491

**Review by pomek — APPROVED** [review]
R+ in terms of pkg.json and lock changes.
